### PR TITLE
DAG: Replace bitwidth with type in suffix in atomic tablegen ops

### PR DESCRIPTION
**Pull request by arsenm**
For FP atomics involving bfloat vs. half, we need to distinguish the type and not rely on the bitwidth alone. For my purposes, an alternative would be to relax the atomic predicate MemoryVT pattern check with a memory size only check. Since there are no extending operations involved, the pattern value check should be unambiguous.

For some reason when using the _32 variants for atomicrmw fadd, I was able to select v2f16 but v2bf16 would fail.

Changes mostly done with sed, e.g.
sed -E -i -r 's/atomic_load_(add|swap|sub|and|clr|or|xor|nand|min|max|umin|umax|swap)_([0-9]+)/atomic_load_\1_i\2/' llvm/lib/Target/*/*.td

### CODE REVIEW

**Review by RKSimon — COMMENTED** [review]
x86 LGTM (I've trimmed out the trailing whitespace in 46d94bd0ad0c22a686ea71f6e7d0494f74c22f1a)

**Review by ritter-x2a — COMMENTED** [review]
The two occurrences of `swap` in the sed command don't match with the definitions in the code ([here](https://github.com/llvm/llvm-project/blob/2e4c06780c14a9c8445129bdbcd15ceaf25e7ecd/llvm/include/llvm/Target/TargetSelectionDAG.td#L1739-L1751)). In AMDGPU, I didn't see a case where that was a problem, but it might be elsewhere.
Other than that and my inline comment, AMDGPU looks fine.

File: llvm/include/llvm/Target/TargetSelectionDAG.td [hunk 1]
```diff
@@ -1680,60 +1680,38 @@ multiclass ternary_atomic_op_ord {
...
-    let MemoryVT = !if(IsInt, i32, f32);
-  }
-  def _64 : PatFrag<(ops node:$ptr, node:$val),
-                    (atomic_op node:$ptr, node:$val)> {
-    let IsAtomic = true;
-    let MemoryVT = !if(IsInt, i64, f64);
+multiclass binary_atomic_op<SDNode atomic_op> {
+  foreach vt = [ i8, i16, i32, i64 ] in {
+    def _#vt : PatFrag<(ops node:$ptr, node:$val),
+                     (atomic_op  node:$ptr, node:$val)> {
```
Comment (jrtc27) [inline llvm/include/llvm/Target/TargetSelectionDAG.td:1686]: Indentation looks off for all your PatFrags

**Comment by jrtc27** [comment]
This seems fine to me

**Comment by jrtc27** [comment]
Though I'll note that it would be nice to have unsuffixed ops, like we do for most ops, so you can just let TableGen's inference figure it out when you're not dealing with values less than the register size.

**Comment by arsenm** [comment]
> Though I'll note that it would be nice to have unsuffixed ops, like we do for most ops, so you can just let TableGen's inference figure it out when you're not dealing with values less than the register size.

I'm not sure why atomics ended up doing this, but I'd guess it's related to the mistake of having ATOMIC_LOAD be a separate opcode from LOAD. ATOMIC_LOAD/STORE never really consistently handled the extload/truncstore cases, which was then copied to the atomicrmw

**Comment by arsenm** [comment]
> This seems fine to me

Can you mark as approved?

**Review by RKSimon — APPROVED** [review]

**Pull request merged**